File: day6.c
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#define A(i, j) a[(i) & 255][(j) & 255]
char a[256][256], b[256][256];
int n;

const int di[] = {-1, 0, 1, 0 };
const int dj[] = { 0, 1, 0,-1 };

int walk(int si, int sj)
{
    int i = si, j = sj;
    int d = 0;
    int steps = 0;
    memcpy(b, a, sizeof(a));
    while (A(i, j)) {
        if (A(i, j) == '.') {
            A(i, j) = 'X' + d;
            steps++;
        } else if (A(i, j) == 'X' + d) {
            steps = -steps;
            break;
        }
        int ni = i + di[d], nj = j + dj[d];
        if (A(ni, nj) == '#') {
            d = (d + 1) & 3;
        } else {
            i = ni, j = nj;
        }
    }
    memcpy(a, b, sizeof(a));
    return steps;
}

int main()
{
    FILE* file = fopen("day6.txt", "r");
    int sum = 0, sum2 = 0;

    while (1 == fscanf(file, "%s", &A(n, 0))) n++;

    int si, sj;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (A(i, j) == '^') {
                A(i, j) = '.';
                si = i, sj = j;
                break;
            }
        }
    }

    sum = walk(si, sj);

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (A(i, j) == '.') {
                A(i, j) = '#';
                int steps = walk(si, sj);
                if (steps < 0) sum2++;
                A(i, j) = '.';
            }
        }
    }

    printf("%d %d\n", sum, sum2);
}

File: day6.lua
#!/usr/bin/env lua

local mt = { __index = function(t, k) t[k] = {} return t[k] end }

local byteDot = string.byte '.'
local byteX = string.byte 'X'
local byteHash = string.byte '#'
local byteCaret = string.byte '^'

local di = {-1, 0, 1, 0 }
local dj = { 0, 1, 0,-1 }

local function walk(a, si, sj)
    local i, j = si, sj
    local d = 1
    local steps = 0
    local b = {}
    setmetatable(b, mt)
    while a[i][j] do
        if not b[i][j] then
            b[i][j] = byteX + d
            steps = steps + 1
        elseif b[i][j] == byteX + d then
            steps = -steps
            break
        end
        local ni, nj = i + di[d], j + dj[d]
        if a[ni][nj] == byteHash then
            d = (d & 3) + 1
        else
            i, j = ni, nj
        end
    end
    return steps
end

local file = io.open("day6.txt")
local sum, sum2 = 0, 0
local a = {}
setmetatable(a, mt)

for line in file:lines() do
    table.insert(a, { line:byte(1, #line) })
end

local si, sj
for i = 1, #a do
    for j = 1, #a do
        if a[i][j] == byteCaret then
            a[i][j] = byteDot
            si, sj = i, j
            break
        end
    end
end

sum = walk(a, si, sj)

for i = 1, #a do
    for j = 1, #a do
        if a[i][j] == byteDot then
            a[i][j] = byteHash
            local steps = walk(a, si, sj)
            if steps < 0 then sum2 = sum2 + 1 end
            a[i][j] = byteDot
        end
    end
end

print(sum, sum2)
